File: Stochastic_Evolutionary_Game/Stochastic_Game_Nowak_Code/Public_Goods_Game/Matlab_Version/GetEvolData.m
function [coop, freq, Data] = GetEvolData()
    
    % [coop, freq, Data] = GetEvolData();
    % Create the data for Figure 2b by calling the subroutine EvolProc.m
    % that simulates the dynamics of the stochasitc game, and of the two
    % corresponding repeated games.
    % coop ... matrix that contains the average cooperation rate in
    % population
    % for each timestep and each of the three scenarios
    % freq ... matrix that contains the average abundance of each strategy
    % Data ... stores the parameters used for the simulations
    
    %% Setting up the objects and defining the parameters
    grSize = 4; beta = 100; r2 = 1.2; r1 = 1.6; c = 1; nGen = 10^4; nIt = 100; % Parameters in Fig. 2b
    coopS = zeros(1, nGen); coop1 = coopS; coop2 = coopS; % Vectors that store the cooperation rates for each scenario in each round
    freqS = zeros(1, 2^(2*grSize)); freq1 = freqS; freq2 = freqS; % Vectors that store the average frequency of each memory-1 strategy
    qS = [1, zeros(1, grSize)]; q1 = ones(1, grSize+1); q2 = zeros(1, grSize+1); % Defining the transition of the three scenarios
    
    for t = 1:nIt % Running nIt independent simulations of the dynamics and averaging over the outcomes
        t;
        [coS, freS] = EvolProc(qS, r1, r2, c, beta, nGen); coopS = (t-1)/t*coopS + coS/t; freqS = (t-1)/t*freqS + freS/t;
        [co1, fre1] = EvolProc(q1, r1, r1, c, beta, nGen); coop1 = (t-1)/t*coop1 + co1/t; freq1 = (t-1)/t*freq1 + fre1/t;
        [co2, fre2] = EvolProc(q2, r2, r2, c, beta, nGen); coop2 = (t-1)/t*coop2 + co2/t; freq2 = (t-1)/t*freq2 + fre2/t;
    end
    
% Creating the output
coop = [coopS; coop1; coop2];
freq = [freqS; freq1; freq2];
Data = ['r1=', num2str(r1), '; r2=', num2str(r2), '; c=', num2str(c), '; beta=', num2str(beta), '; nIt=', num2str(nIt)]; 
end

File: Stochastic_Evolutionary_Game/Stochastic_Game_Nowak_Code/Public_Goods_Game/Python_Version/GetEvolData.py
import numpy as np
from EvolProc import EvolProc
import argparse
import datetime

parser = argparse.ArgumentParser()
parser.add_argument("-r1", "--r1", type=float, default=1.6, help="The parameter of state 1")
parser.add_argument("-r2", "--r2", type=float, default=1.2, help="The parameter of state 2")
args = parser.parse_args()

def GetEvolData():
    """
    Creates the data for the dynamics of cooperation rate with time
    Simulates the dynamics of the stochastic game, and of the two corresponding repeated game
    :return:
    coop: matrix that contains the average cooperation rate in population
    for each timestep and each of the three scenarios
    freq: matrix that contains the average abundance of each strategy
    Data: stores the parameters used for the simulations
    """

    # Setting up the objects and definning the parameters
    grSize = 4; beta = 100; r2 = 1.2; c = 1; r1 = 1.6; nGen = 10**4; nIt = 100; # Parameters in Fig. 2b
    coopS = np.zeros((1, nGen)).flatten(); coop1 = np.zeros((1, nGen)).flatten(); coop2 = np.zeros((1, nGen)).flatten()
    freqS = np.zeros((1, 2**(2*grSize))).flatten(); freq1 = np.zeros((1, 2**(2*grSize))).flatten(); freq2 = np.zeros((1, 2**(2*grSize))).flatten()
    qS = np.append(np.array([1]), np.zeros((grSize))); q1 = np.ones((1, grSize+1)).flatten(); q2 = np.zeros((1, grSize+1)).flatten() # Defining the transitions of the three scenarios

    for i in range(nIt):
        print (i)
        starttime = datetime.datetime.now()
        (coop, freq) = EvolProc(qS, r1, r2, c, beta, nGen); coopS = i/(i+1)*coopS+1/(i+1)*coop; freqS = i/(i+1)*freqS+1/(i+1)*freq
        endtimeS = datetime.datetime.now()
        print ("Complete State s", (endtimeS-starttime).seconds)
        (coop, freq) = EvolProc(q1, r1, r2, c, beta, nGen); coop1 = i/(i+1)*coop1+1/(i+1)*coop; freq1 = i/(i+1)*freq1+1/(i+1)*freq
        endtime1 = datetime.datetime.now()
        print ("Complete State 1", (endtime1-starttime).seconds)
        (coop, freq) = EvolProc(q2, r1, r2, c, beta, nGen); coop2 = i/(i+1)*coop2+1/(i+1)*coop; freq2 = i/(i+1)*freq2+1/(i+1)*freq
        endtime2 = datetime.datetime.now()
        print ("Complete State 2", (endtime2-starttime).seconds)

    # Creating the output
    coop = np.array([coopS, coop1, coop2])
    freq = np.array([freqS, freq1, freq2])
    return (coop, freq)

def writefile(filename, dataname):
    f = open(filename, 'w')
    shape = dataname.shape
    for i in range(shape[0]):
        for j in range(shape[1]):
            if j < (shape[1] - 1):
                f.write(str(dataname[i][j]) + ',')
            else:
                f.write(str(dataname[i][j]))
        f.write('\n')
    f.close()

(coopRe, freRe) = GetEvolData()
writefile("Pub_Coop_Time.txt", coopRe)

